nerfstudio/model_components/losses.py

```diff
@@ -308,3 +308,234 @@ def depth_loss(
 
     raise NotImplementedError("Provided depth loss type not implemented.")
+
+
+def monosdf_normal_loss(normal_pred: torch.Tensor, normal_gt: torch.Tensor) -> torch.Tensor:
+    """normal consistency loss as monosdf https://niujinshuchong.github.io/monosdf/
+    Args:
+        normal_pred (torch.Tensor): volume rendered normal
+        normal_gt (torch.Tensor): monocular normal
+    """
+    normal_gt = torch.nn.functional.normalize(normal_gt, p=2, dim=-1)
+    normal_pred = torch.nn.functional.normalize(normal_pred, p=2, dim=-1)
+    l1 = torch.abs(normal_pred - normal_gt).sum(dim=-1).mean()
+    cos = (1.0 - torch.sum(normal_pred * normal_gt, dim=-1)).mean()
+    return l1 + cos
+
+
+# copy from MiDaS https://github.com/EPFL-VILAB/omnidata/blob/main/omnidata_tools/torch/losses/midas_loss.py#L8
+def compute_scale_and_shift(prediction: torch.Tensor, target: torch.Tensor, mask: torch.Tensor):
+    """
+    https://github.com/EPFL-VILAB/omnidata/blob/318a75569934737e67902f903531324d1f48ae8f/omnidata_tools/torch/losses/midas_loss.py#L8
+    Args:
+        prediction: predicted depth map
+        target: ground truth depth map
+        mask: mask of valid pixels
+    Returns:
+        scale and shift for depth prediction
+    """
+    # system matrix: A = [[a_00, a_01], [a_10, a_11]]
+    a_00 = torch.sum(mask * prediction * prediction, (1, 2))
+    a_01 = torch.sum(mask * prediction, (1, 2))
+    a_11 = torch.sum(mask, (1, 2))
+
+    # right hand side: b = [b_0, b_1]
+    b_0 = torch.sum(mask * prediction * target, (1, 2))
+    b_1 = torch.sum(mask * target, (1, 2))
+
+    # solution: x = A^-1 . b = [[a_11, -a_01], [-a_10, a_00]] / (a_00 * a_11 - a_01 * a_10) . b
+    x_0 = torch.zeros_like(b_0)
+    x_1 = torch.zeros_like(b_1)
+
+    det = a_00 * a_11 - a_01 * a_01
+    valid = det.nonzero()
+
+    x_0[valid] = (a_11[valid] * b_0[valid] - a_01[valid] * b_1[valid]) / det[valid]
+    x_1[valid] = (-a_01[valid] * b_0[valid] + a_00[valid] * b_1[valid]) / det[valid]
+
+    return x_0, x_1
+
+
+def reduction_batch_based(image_loss: torch.Tensor, M: torch.Tensor):
+    """
+    average of all valid pixels of the batch
+    Args:
+        image_loss: loss of each image
+        M: number of valid pixels of each image
+    Returns:
+        average of all valid pixels of the batch
+    """
+
+    # avoid division by 0 (if sum(M) = sum(sum(mask)) = 0: sum(image_loss) = 0)
+    divisor = torch.sum(M)
+    reduction = 0 if divisor == 0 else torch.sum(image_loss) / divisor
+
+    return reduction
+
+
+def reduction_image_based(image_loss, M):
+    """
+    mean of average of valid pixels of an image
+    Args:
+        image_loss: loss of each image
+        M: number of valid pixels of each image
+    Returns:
+        mean of average of valid pixels of an image
+    """
+    # avoid division by 0 (if M = sum(mask) = 0: image_loss = 0)
+    valid = M.nonzero()
+
+    image_loss[valid] = image_loss[valid] / M[valid]
+
+    return torch.mean(image_loss)
```

Comment:
I feel like the logic here can be build directly into the functions, it seems pretty minimal in practice.